tests/testserver/server.py

```diff
@@ -79,13 +81,33 @@ def _create_socket_and_bind(self):
         sock.listen(0)
         return sock
 
+    def _close_server_sock_ignore_errors(self):
+        try:
+            self.server_sock.close()
+        except IOError:
+            pass
+
     def _handle_requests(self, server_sock):
-        for _ in range(self.requests_to_handle):
-            sock = server_sock.accept()[0]
+        for i in range(self.requests_to_handle):
+            sock = self._accept_connection(server_sock)
+            if not sock:
+                break
+
             handler_result = self.handler(sock)
 
             self.handler_results.append(handler_result)
 
+    def _accept_connection(self, server_sock):
+        try:
+            ready, _, _ = select.select([server_sock], [], [])
+            if not ready:
+                return None
+
+            return server_sock.accept()[0]
+        except (select.error, socket.error) as e:
+            print("Failed to accept connection: {0}".format(e))
+            return None
```

Comment:
I wanted to be able to interrupt the server thread if an exception occurred in the main thread. In the instance that this fails, something is going wrong elsewhere, and I feel that the stack traces from the failed select/accept will just obscure the real failure.